PROTO_tests/tests/numeric_test.py

```diff
@@ -604,3 +604,96 @@ def test_hash(self):
+                    nd_arry = ia.astype(dtype3)
+                    ak_arry = ak.array(nd_arry)
+                    assert np.allclose(np.clip(nd_arry, lo, hi), ak.clip(ak_arry, lo, hi).to_ndarray())
+
+        # test with None for lower limit
+
+        lo = None
+        for dtype1 in dtypes:
+            for dtype2 in dtypes:
+                for dtype3 in dtypes:
+                    hi = np.dtype(dtype1).type(ihi)
+                    nd_arry = ia.astype(dtype3)
+                    ak_arry = ak.array(nd_arry)
+                    assert np.allclose(np.clip(nd_arry, lo, hi), ak.clip(ak_arry, lo, hi).to_ndarray())
+
+        # test with None for upper limit
+
+        hi = None
+        for dtype1 in dtypes:
+            for dtype2 in dtypes:
+                for dtype3 in dtypes:
+                    lo = np.dtype(dtype2).type(ilo)
+                    nd_arry = ia.astype(dtype3)
+                    ak_arry = ak.array(nd_arry)
+                    assert np.allclose(np.clip(nd_arry, lo, hi), ak.clip(ak_arry, lo, hi).to_ndarray())
+
+        # test with arrays for lo
+
+        ilo = np.full(ia.shape, ilo)
+        for dtype1 in dtypes:
+            for dtype2 in dtypes:
+                for dtype3 in dtypes:
+                    hi = np.dtype(dtype1).type(ihi)
+                    nd_lo = ilo.astype(dtype2)
+                    nd_arry = ia.astype(dtype3)
+                    ak_lo = ak.array(nd_lo)
+                    ak_arry = ak.array(nd_arry)
+                    assert np.allclose(
+                        np.clip(nd_arry, nd_lo, hi),
+                        ak.clip(ak_arry, ak_lo, hi).to_ndarray(),
+                    )
+
+        # test with arrays for hi
+
+        ilo = 25
+        ihi = np.full(ia.shape, ihi)
+        for dtype1 in dtypes:
+            for dtype2 in dtypes:
+                for dtype3 in dtypes:
+                    lo = np.dtype(dtype2).type(ilo)
+                    nd_hi = ihi.astype(dtype1)
+                    nd_arry = ia.astype(dtype3)
+                    ak_hi = ak.array(nd_hi)
+                    ak_arry = ak.array(nd_arry)
+                    assert np.allclose(
+                        np.clip(nd_arry, lo, nd_hi),
+                        ak.clip(ak_arry, lo, ak_hi).to_ndarray(),
+                    )
+
+        # test with arrays for both
+
+        ilo = np.full(ia.shape, ilo)
+        for dtype1 in dtypes:
+            for dtype2 in dtypes:
+                for dtype3 in dtypes:
+                    nd_hi = ihi.astype(dtype1)
+                    nd_lo = ilo.astype(dtype2)
+                    nd_arry = ia.astype(dtype3)
+                    ak_lo = ak.array(nd_lo)
+                    ak_hi = ak.array(nd_hi)
+                    ak_arry = ak.array(nd_arry)
+                    assert np.allclose(
+                        np.clip(nd_arry, nd_lo, nd_hi),
+                        ak.clip(ak_arry, ak_lo, ak_hi).to_ndarray(),
+                    )
```

Comment:
Overall this all looks great. The one think that I would recommend is taking more advantage of the parameterization that PROTO_test offers. This may be able to reduce the amount of code that you have in these sections. You do not have very many parameters, so there may not be as much of an improvement as there are other places.

The other suggestion is to combine the logic of the singleton cases and the array cases into a singe for loop. If you generate the `ilo` and `ihi` arrays, you can then use `ilo[0]` and `ihi[0]` and combine several sections by doing the singleton and array tests in the same for loop. Similar things were done in other places in `numeric_test.py` in places like `sqrt`, `power`, and `arctan2`.

Reply:
Updated tests look good